brainwidemap/encoding/utils.py

```diff
@@ -0,0 +1,295 @@
+    ret_wheel : bool, optional
+        Whether to return the time-resampled wheel velocity trace, by default False
+    ret_abswheel : bool, optional
+        Whether to return the time-resampled absolute wheel velocity trace, by default False
+    wheel_binsize : float, optional
+        Time bins to resample wheel velocity to, by default 0.02
+    addtl_types : list, optional
+        List of additional types from an ONE trials object to include in the dataframe. Must be
+        valid keys to the dict produced by one.load_object(eid, 'trials'), by default empty.
+    trials_mask : list, optional
+        List of trial indices to include in the dataframe. If None, all trials will be included.
+
+    Returns
+    -------
+    pandas.DataFrame
+        Dataframe with trial-wise information. Indices are the actual trial order in the original
+        data, preserved even if some trials do not meet the maxlen criterion. As a result will not
+        have a monotonic index. Has special columns trial_start and trial_end which define start
+        and end times via t_before and t_after
+    """
+    if not one:
+        one = ONE()
+
+    if ret_wheel and ret_abswheel:
+        raise ValueError("ret_wheel and ret_abswheel cannot both be true.")
+
+    # Define which datatypes we want to pull out
+    trialstypes = [
+        "choice",
+        "probabilityLeft",
+        "feedbackType",
+        "feedback_times",
+        "contrastLeft",
+        "contrastRight",
+        "goCue_times",
+        "stimOn_times",
+    ]
+    trialstypes.extend(addtl_types)
+
+    loader = SessionLoader(one=one, eid=eid)
+    loader.load_session_data(pose=False, motion_energy=False, pupil=False, wheel=True)
+    loader.load_wheel(fs=1000, corner_frequency=70, order=8)
+
+    trials = loader.trials
+    starttimes = trials.stimOn_times
+    endtimes = trials.feedback_times
+    trialsdf = trials[trialstypes]
+    if trials_mask is not None:
+        trialsdf = trialsdf.loc[trials_mask]
+    trialsdf["trial_start"] = trialsdf["stimOn_times"] - t_before
+    trialsdf["trial_end"] = trialsdf["feedback_times"] + t_after
+    tdiffs = trialsdf["trial_end"] - np.roll(trialsdf["trial_start"], -1)
+    if np.any(tdiffs.iloc[:-1] > 0):
+        logging.warning(
+            f"{sum(tdiffs[:-1] > 0)} trials overlapping due to t_before and t_after "
+            "values. Try reducing one or both!"
+        )
+    if not ret_wheel and not ret_abswheel:
+        return trialsdf
+
+    wheel = one.load_object(eid, "wheel", collection="alf")
+    whlpos, whlt = wheel.position, wheel.timestamps
+    starttimes = trialsdf["trial_start"]
+    endtimes = trialsdf["trial_end"]
+    wh_endlast = 0
+    trials = []
+    for (start, end) in np.vstack((starttimes, endtimes)).T:
+        wh_startind = np.searchsorted(whlt[wh_endlast:], start) + wh_endlast
+        wh_endind = np.searchsorted(whlt[wh_endlast:], end, side="right") + wh_endlast + 4
```

Comment:
Basically it's a pad to make sure the sampled start and end inds will go a little beyond the end of the trial window.

This is because later on when I resample the wheel trace I do so conservatively. If e.g. I don't have a value before and after a desired resampling time the resampling will place a NaN there. This makes sure that even at the end of the trial I have several bins to work with during the resampling.

Either way it's cut down after the TimeSeries resampling, and 4 bins is only 80ms or so using the default binwidth so unlikely to get the next trial in the pad.

Admittedly it's not great to use a magic number like this but it works for now